geopandas/geodataframe.py

```diff
@@ -298,7 +298,7 @@ def set_geometry(self, col, drop=False, inplace=False, crs=None):
         geo_column_name = self._geometry_column_name
         if isinstance(col, (Series, list, np.ndarray, GeometryArray)):
             level = col
-        elif hasattr(col, "ndim") and col.ndim != 1:
+        elif hasattr(col, "ndim") and col.ndim > 1:
```

Comment:
In _theory_ the user could pass a 0-dim array, and that should still raise?

Reply:
I will have to think on this a bit more. I made that change because concat(axis=1) of two geoseries will generate default column names [0,1], but this is actually [np.int64(0), np.int64(1)], which have ndim zero (I can't remember if that happens all the time, but it's why I added the test `test_set_geometry_np_int` below). I'm not sure what another 0-dim array might be? short of an arbitrary class with that attribute? 
test_set_geometry_np_int